### PR TITLE
Add execution method which accepts json data

### PR DESCRIPTION
**Pull request by DriesVanbilloen**
This makes it possible to execute the migration with a data json object via code 
(Same functionality as in the groovy console)

Also added new filter based on rootpaths

### CODE REVIEW

**Comment by royteeuwen** [comment]
@gruberrolandvaltech  could you tell us who the new maintainer is so that we can get this merged?

**Comment by nhirrle** [comment]
Hi @DriesVanbilloen 

Thanks a lot for your contribution!

Some minor comments:
When building I get a few complaints from the bnd plugin. Could you increase the version numbers in the package-info:
de.valtech.aecu.api.groovy.console.bindings 4.8.0 --> 4.9.0
de.valtech.aecu.api.service 3.0.0 --> 3.1.0

and in the pom bundle version 6.2.0 to 6.3.0

Could you add an update within the Readme.md for the data option? As this is a neat feature I think it's worth mentioning it :) 
Also I would like to understand how it works, can we use this within Groovyconsole as well?

File: api/src/main/java/de/valtech/aecu/api/groovy/console/bindings/filters/FilterByRootPaths.java
Comment (nhirrle) [inline api/src/main/java/de/valtech/aecu/api/groovy/console/bindings/filters/FilterByRootPaths.java:20]: as this is a rootpath contains is not correct. can you change the condition to 
```
currentPath.equals(rootPath) || (currentPath + "/" ).startsWith(rootPath) 
```

File: api/src/main/java/de/valtech/aecu/api/groovy/console/bindings/filters/FilterByRootPaths.java
Comment (nhirrle) [inline api/src/main/java/de/valtech/aecu/api/groovy/console/bindings/filters/FilterByRootPaths.java:8]: please add javadoc and rename to filterByRootNodePaths

File: api/src/main/java/de/valtech/aecu/api/groovy/console/bindings/filters/FilterByRootPaths.java
Comment (nhirrle) [inline api/src/main/java/de/valtech/aecu/api/groovy/console/bindings/filters/FilterByRootPaths.java:12]: also here please add javadoc

**Comment by DriesVanbilloen** [comment]
The data parameter is actually added to mimic the behaviour of the groovyconsole.
This data (json object) can now be used within the java api to inject that data object in the scripts context :)

I will resolve the remarks today!

**Comment by DriesVanbilloen** [comment]
To make the data usage more clear, I will add a jmx method so it can also be explained in the docs!

**Comment by DriesVanbilloen** [comment]
> Hi @DriesVanbilloen
> 
> Thanks a lot for your contribution!
> 
> Some minor comments: When building I get a few complaints from the bnd plugin. Could you increase the version numbers in the package-info: de.valtech.aecu.api.groovy.console.bindings 4.8.0 --> 4.9.0 de.valtech.aecu.api.service 3.0.0 --> 3.1.0
> 
> and in the pom bundle version 6.2.0 to 6.3.0
> 
> Could you add an update within the Readme.md for the data option? As this is a neat feature I think it's worth mentioning it :) Also I would like to understand how it works, can we use this within Groovyconsole as well?

What is the pom bundle version exactly?

**Comment by nhirrle** [comment]
> To make the data usage more clear, I will add a jmx method so it can also be explained in the docs!

This would be appreciated, great idea.

**Comment by nhirrle** [comment]
> > Hi @DriesVanbilloen
> > Thanks a lot for your contribution!
> > Some minor comments: When building I get a few complaints from the bnd plugin. Could you increase the version numbers in the package-info: de.valtech.aecu.api.groovy.console.bindings 4.8.0 --> 4.9.0 de.valtech.aecu.api.service 3.0.0 --> 3.1.0
> > and in the pom bundle version 6.2.0 to 6.3.0
> > Could you add an update within the Readme.md for the data option? As this is a neat feature I think it's worth mentioning it :) Also I would like to understand how it works, can we use this within Groovyconsole as well?
> 
> What is the pom bundle version exactly?

Please have a look when running `mvn clean install` 
https://github.com/valtech/aem-easy-content-upgrade/actions/runs/4122005488/jobs/7275025541#step:6:23524
```
 Error:  Baseline mismatch for package de.valtech.aecu.api.groovy.console.bindings, MINOR change. Current is 4.8.0, repo is 4.8.0, suggest 4.9.0 or -
Error:  Baseline mismatch for package de.valtech.aecu.api.groovy.console.bindings.filters, MINOR change. Current is 3.1.0, repo is 3.1.0, suggest 3.2.0 or -
Error:  Baseline mismatch for package de.valtech.aecu.api.service, MINOR change. Current is 3.0.0, repo is 3.0.0, suggest 3.1.0 or -
Error:  The bundle version change (6.2.0 to 6.2.1) is too low, the new version must be at least 6.3.0
``` 

versions 6.2.1-SNAPSHOT should be changed to 6.3.0-SNAPSHOT in the different pom.xml's. Please also change it in sonar.projectVersion in sonar-project.properties file

**Comment by royteeuwen** [comment]
@nhirrle both are fine by me. We already have an internal fork for as long as there is no public release :)! What is easiest for you

**Comment by nhirrle** [comment]
> @nhirrle both are fine by me. We already have an internal fork for as long as there is no public release :)! What is easiest for you

OK good, in that case I just wait until you are done :)

**Comment by nhirrle** [comment]
`[ERROR] Baseline mismatch for package de.valtech.aecu.api.groovy.console.bindings.filters, MINOR change. Current is 3.1.0, repo is 3.1.0, suggest 3.2.0`
Please update version in de/valtech/aecu/api/groovy/console/bindings/filters/package-info.java from 3.1.0 to 3.2.0

File: core/src/main/java/de/valtech/aecu/core/jmx/AecuServiceMBean.java
Comment (nhirrle) [inline core/src/main/java/de/valtech/aecu/core/jmx/AecuServiceMBean.java:96]: in the description it should be added that json data is expected

File: api/src/main/java/de/valtech/aecu/api/service/AecuService.java
Comment (nhirrle) [inline api/src/main/java/de/valtech/aecu/api/service/AecuService.java:151]: the javadoc is wrong as we cannot pass in the position. also in the implementation we have /apps hardcoded but scripts could also be stored elsewhere, e.g. within /conf . 

Suggestion: We remove the executeAllScripts method and implement the JMX interface for method `execute(String path, String data)`

Comment (DriesVanbilloen) [reply api/src/main/java/de/valtech/aecu/api/service/AecuService.java:151]: > the javadoc is wrong as we cannot pass in the position. also in the implementation we have /apps hardcoded but scripts could also be stored elsewhere, e.g. within /conf .
> 
> Suggestion: We remove the executeAllScripts method and implement the JMX interface for method `execute(String path, String data)`

I get that, but that is not the point of this method. The execute(String path, String data) is only for 1 groovyscript, while  the execute all scripts is for all script. 

Counter suggestion: executeScripts(String[] scriptRoots, String data)

What do you think?

Comment (nhirrle) [reply api/src/main/java/de/valtech/aecu/api/service/AecuService.java:151]: > > the javadoc is wrong as we cannot pass in the position. also in the implementation we have /apps hardcoded but scripts could also be stored elsewhere, e.g. within /conf .
> > Suggestion: We remove the executeAllScripts method and implement the JMX interface for method `execute(String path, String data)`
> 
> I get that, but that is not the point of this method. The execute(String path, String data) is only for 1 groovyscript, while the execute all scripts is for all script.
> 
> Counter suggestion: executeScripts(String[] scriptRoots, String data)
> 
> What do you think?

Hi @DriesVanbilloen 
Mmh, I just tested and it is enough to pass in a root path 
grafik
Internally findCandidates is used https://github.com/valtech/aem-easy-content-upgrade/blob/master/core/src/main/java/de/valtech/aecu/core/service/AecuServiceImpl.java#L111

Comment (DriesVanbilloen) [reply api/src/main/java/de/valtech/aecu/api/service/AecuService.java:151]: Mmh, missed that. Ok I will apply your suggestions

**Comment by DriesVanbilloen** [comment]
@nhirrle Checks have passed and the suggestions are applied. Are we waiting for a certain approval?

**Comment by nhirrle** [comment]
> @nhirrle Checks have passed and the suggestions are applied. Are we waiting for a certain approval?

Hi @DriesVanbilloen 
Could you try to push again? I don't see the updated JMX method in Github yet. (https://github.com/valtech/aem-easy-content-upgrade/pull/204#discussion_r1113290816 )

**Review by nhirrle — APPROVED** [review]

**Pull request merged**
